### ISSUE TITLE
Implement "both hands" description

### ISSUE THREAD

**Issue by zepumph** [phetsims/ratio-and-proportion#173]
This issue combines much of https://github.com/phetsims/ratio-and-proportion/issues/172, https://github.com/phetsims/ratio-and-proportion/issues/161 and https://github.com/phetsims/ratio-and-proportion/issues/157 (mostly the design coming from 157).

During design meeting today, we decided on the following direction for both hands:

On focus, we want to state the distance between each hand, as explained in https://github.com/phetsims/ratio-and-proportion/issues/157#issuecomment-686741206. We also want this behavior whenever either hand changes.

Then on "interaction end" we want a context alert like "Extremely close to challenge ratio, {{left hand near bottom of Play Area, right just below middle}}." The first part is the same qualitative ratio description used elsewhere in the sim. The second part was taken from https://github.com/phetsims/ratio-and-proportion/issues/157#issuecomment-686777897.

This is my top priority.

**Comment by zepumph** [comment]
In design meeting today, we discussed implementing this with aria-live, at least in part, but after investigation over in https://github.com/phetsims/a11y-research/issues/159, I think that the both hands on focus/change description should be attempted with aria-describedby first. I'll see how it goes.

**Comment by zepumph** [comment]
More notes from the meeting today:
```
Both hands stuff
Use this on individual hands when the ratio is locked.
* Hands {{somewhat close to each other}}, Both Hands, movable <- on focus
* Hands {{somewhat close to each other}} <- value/position change
* extremely close to challenge ratio, {{Left hand near bottom of Play Area, right just below middle}}. <- after a bit, "end of interaction"```

**Comment by zepumph** [comment]
Above implemented the distance-related both hands description using aria-describedby, because https://github.com/phetsims/a11y-research/issues/159 went pretty well. I think it is a bit less responsive than aria-valuetext, but hopefully it will still be acceptable. 

Next I will rework the context alert to include ratio qualitative and hand positions.

**Comment by zepumph** [comment]
[Above](https://github.com/phetsims/ratio-and-proportion/commit/c9fb5989e99078cafe6f0eece092325af466640b) implemented the new ratio/position alert. Next I will look into converting single hand behavior to this when the ratio is locked.

**Comment by zepumph** [comment]
Thoughts for review:

- [ ] Currently distance says something like "Hands somewhat close to each other." but I think we should lose the capitalization and period, because we are more mimicking aria-valuetext conventions. Plus when ratio is locked from work over in https://github.com/phetsims/ratio-and-proportion/issues/172 that description actually IS in aria-valuetext. Do you agree that we should make this change?

Ready for review. Assigning to multiple people. Dev link coming soon.

**Comment by zepumph** [comment]
Dev version over in https://github.com/phetsims/ratio-and-proportion/issues/173

**Comment by Matthew-Moore240** [comment]
In play area is repeated twice in the aria live text for both hands.

**Comment by zepumph** [comment]
When I use the "both hands" interaction, I get alerts like this:

>Not so close to challenge ratio. Left hand on second tick mark of Play area, right hand just above sixth tick mark
Somewhat close to challenge ratio. Left hand on second tick mark of Play area, right hand halfway past fifth tick mark

That seems right to me.

Then when I look at the alerts while using a single hand while the ratio is locked, alerts look like:

>At challenge ratio. Left hand somewhat near bottom of Play Area of Play area, right hand just below middle
At challenge ratio. Left hand near bottom of Play Area of Play area, right hand just below middle

I will take a look at that.

**Comment by zepumph** [comment]
My above assessment was a bit wrong. Actually I was making sure to always add "of Play Area" to the left hand description, but it didn't duplicate for (semi-)quantitative. Above I removed the extra "of Play Area", and now we are also never getting that for any type of quantitative descriptions.

Good catch @Matthew-Moore240!

**Comment by zepumph** [comment]
Let's go with https://phet-dev.colorado.edu/html/ratio-and-proportion/1.0.0-dev.62/phet/ now.

**Comment by Matthew-Moore240** [comment]
This looks good to me for a first run through. I do think some of the regions are off and the timer is just a hair to long. Great job @zepumph

**Comment by terracoda** [comment]
@Matthew-Moore240, what about putting the distance between the hands description in the first part of the description? I think this might be a bit more natural sounding. And the structure may be more parallel to the actual interactive pieces (two bullets = two hands).
>> Currently, hands are at the challenge 1 ratio with
>> - Left hand just below middle of Play Area
>> - Right hand near top of Play Area
>> - Hands are far from each other.

Currently, hands are {{at}} the challenge {{1}} ratio **and** {{far from}} each other:
- Left hand {{just below middle of}} Play Area
- Right hand {{near top of}} Play Area.

Currently, hands are {{at}} the challenge {{1}} ratio **and** {{close to eachother}}:
- Left hand {{near bottom of}} Play Area
- Right hand {{just below middle of}} Play Area.

Currently, hands are {{extremely close to}} the challenge {{1}} ratio **and** {{not so close to}} each other:
- Left hand {{somewhat near bottom}} of Play Area
- Right hand {{somewhat near top of}} Play Area.

**Tick Marks (plain)**
>> Currently, hands are at the challenge 1 ratio with
>> - Left hand on second tick mark
>> - Right hand on fourth tick mark
>> - Hands are somewhat close to each other.

Currently, hands are {{at}} the challenge {{1}} ratio **and** {{somewhat close to}} each other:
- Left hand {{on}} {{second}} tick mark
- Right hand {{on}} {{fourth}} tick mark.

Currently, hands are {{at}} the challenge 1 ratio **and** {{not so close to}} each other:
- Left hand {{on}} {{third}} tick mark
- Right hand {{on}} sixth}} tick mark.

Currently, hands are {{somewhat close to}} the challenge {{1}} ratio **and** {{very close to}} each other:
- Left hand {{on}} {{third}} tick mark
- Right hand {{on}} {{fourth}} tick mark

**Numbered Tick Marks**
>> Currently, hands are at the challenge 1 ratio with
>> - Left hand on tick mark 2
>> - Right hand on tick mark 4
>> - Hands are somewhat close to each other.

Currently, hands are {{at}} the challenge {{1}} ratio **and** {{somewhat close to}} each other:
- Left hand {{on}} tick mark {{2}}
- Right hand {{on}} tick mark {{4}}.

Currently, hands are at the challenge 1 ratio **and** {{not so close to}} each other:
- Left hand {{on}} tick mark {{3}}
- Right hand on tick mark 6.

Currently, hands are not so close to the challenge 1 ratio **and** are {{far from}} each other:
- Left hand {{on}} tick mark {{2}}
- Right hand {{on}} tick mark {{7}}.

Currently, hands are {{not so close to}} the challenge {{1}} ratio **and** {{very close to}} each other:
- Left hand {{above}} tick mark {{3}}
- Right hand {{halfway past}} tick mark {{4}}.

@Matthew-Moore240, @emily-phet, @zepumph, what do you think of this re-arrangement of the structure?

There are very minor changes to the wording of the pattern. I changes "with" to "and", and put a colon at the end of the initial phrase to hint that something was still coming (the two bullets). If grammatically better, we could put a comma after the firs bullet.

**Comment by zepumph** [comment]
Seems reasonable to me!

**Comment by Matthew-Moore240** [comment]
Yeah I like this approach, it flows nicely.

**Comment by zepumph** [comment]
Sounds great I'll implement

**Comment by zepumph** [comment]
Implemented above, can someone address the thought in https://github.com/phetsims/ratio-and-proportion/issues/173#issuecomment-689176274?

**Comment by Matthew-Moore240** [comment]
@zepumph Sorry I totally missed that question, we should bring it up tomorrow in the meeting I am not 100% sure what we should do there.

**Comment by terracoda** [comment]
@zepumph, will review the alerts with screen reader today.

**Comment by zepumph** [comment]
From today's design meeting, 

- [ ] we found that on VoiceOver, aria-describedby wasn't read on focus, or on change. I will create an a11y-research issue to understand aria-describedby's support better, especially for VoiceOver.

- [ ] RE https://github.com/phetsims/ratio-and-proportion/issues/173#issuecomment-689176274: We decided to lowercase it with no period, even though it is aria-describedby

**Comment by zepumph** [comment]
From investigation over in https://github.com/phetsims/a11y-research/issues/159, I do not think that aria-describedby is the best technology for supplying the responses on the "both hands" interaction. Likely aria-live will be the better solution.

**Comment by zepumph** [comment]
I'm sorry I didn't bring this up during design today. I think it just got lost. @terracoda, I am curious about how the aria-describedby worked (or didn't) in interviews. 

I'm interested in converting this to aria-live + a `p` help-text like element that get's updates as alerts occur. But perhaps there is another way to go about this.

**Comment by terracoda** [comment]
The dynamic description in the screen summary seems to work fine on initial read. I don't think there were many re-reads (actually, may be close to zero re-reads).

I had JAWS users, so the `aria-describedby` worked pretty good during interaction. The learners seemed to get some combination of hand position, distance between hands and proximity to ratio. 

There are still issues around the content - possibly the ordering of the phrases - not scaffolding users to think about ratio more and think less about position/distance. 

@zepumph, we may need a new issue if this one is getting too big?
@Matthew-Moore240 and I need to play with the ordering of the phrases - the details for object responses and details for context responses. Whether they will be delivered solely with `aria live` or with a combination of `aria-describedby` and `aria live` may not be so critical.

**Design-wise**, I am trying to deliver some useful value (some useful brief description of both hands) on focus. Then once interaction starts, that initial information connects with the subsequent object responses and context responses.

The subsequent object & context responses can be delivered with a single technology (`aria live`) or with a combination (`aria-describedby` and `aria live`) if a combination works.

**Re the technology** `aria-describedby`, I found this reference:
 https://a11ysupport.io/tech/aria/aria-describedby_attribute

According to the 2019 test in the above link,  it looks like macOs and iOS support `aria-describedby` on text input elements. I am not sure that will work for moving the hands, I am pretty sure we need role="application" for that.

@zepumph, I took their code:
```html
<label for="target">Example label</label>
<input id="target" aria-describedby="error" type="text">
<div id="error">
error
</div>
```
and adapted it for our context: 
```html
 <label for="both-hands">Both Hands</label>
 <input id="both-hands" aria-describedby="hand-proximity-distance" type="text" role=application aria-roledescription="movable">
 <p id="hand-proximity-distance">somewhat close to each other</p>
<!-- we would still need the proximity to ratio to be delivered with aria-live -->
 ```
I don't know if using an `input type="text"` with role application will make macOs speak the `aria-describedby` on focus and subsequent changes, as well?

@zepumph, if you can get a string of content to be delivered on focus (i.e., an initial object response), we could potentially deliver the rest with`aria live`.

If this all sound crazy. we can discuss further. We are sorting out both the content and how it is delivered, so there is still lots to discuss.

**Comment by zepumph** [comment]
@terracoda I did a little bit of research down the path you outlined above.

This snippet won't work because when we add the  application role, it seems to disable the behavior associated with the `input[text]`. 

```html
 <label for="both-hands">Both Hands</label>
 <input id="both-hands" aria-describedby="hand-proximity-distance" type="text" role=application aria-roledescription="movable">
 <p id="hand-proximity-distance">somewhat close to each other</p>
<!-- we would still need the proximity to ratio to be delivered with aria-live -->
```

When I remove the `role="application"`, then we lose the alt-input behavior, as it behaves like a text input.

I think I'm going to work towards using aria-live for use on change and on focus.

**Comment by zepumph** [comment]
Can we try the above commit, which gives an aria-live alert on focus, and whenever the fitness changes while the "both hands" interaction is being used. 

I only have an iPad, and the interaction doesn't work on that platform, so it should be reviewed for VoiceOver on a Mac book. Over to @terracoda 

Hopefully you can give at least a couple minutes to try this out before we discuss at the description meeting tomorrow.

**Comment by zepumph** [comment]
I brought @terracoda and @Matthew-Moore240 up to speed on the conversion to aria-live today. They will review with this as they continue with the "both hands" design.

**Comment by terracoda** [comment]
@zepumph, the aria-live solution is working. With VoiceOver, I am getting pile-ups of alerts, so we need to optimize that. I am hearing both responses, though, object and context responses, so I think this solution will be good as we work through finalizing the actual strings and their order.

**Comment by terracoda** [comment]
Re pile-ups, I am not sure how you have fixed that in the past, but you have done it before :-)

**Comment by zepumph** [comment]
- [ ] I think we should create a new issue to investigate alert pile ups, but should wait until the description is more refined. Much of the timing may end up being based on the actual length of each alert.

**Comment by terracoda** [comment]
@zepumph, I created #214 to investigate response pile-ups.

**Comment by terracoda** [comment]
I think we should close this issue and continue work over in #207.

**Comment by terracoda** [comment]
@zepumph, please re-open if there are things in here that I have missed.
